US-CBP/GTAS-UI
# Fix/translation fetch

Fixes #672 - Translations not working in dev.

**Update your deps with npm/yarn install if running locally**

Issue was that the translation dependency, i18next, required us to bypass the existing fetch abstraction in servicewrapper, which has the correct header information for all calls, but it was stripping out the additional headers required by our backend in the dev environment. The result was that translation calls made by i18next were not correctly routed to the controller, so frontend never received the codes for translation.

I basically replaced all plugins with my own implementation since they were all causing issues with language resolution, repeat fetch attempts, etc, but that involved updating aspects of the existing context structure, maintaining the state of the last translation load attempt, and handling translation reloads manually.

**To Test**
 1. As admin, go to the language editor page and click the "Live Edit" checkbox
 2. Ensure that the page refreshes and remains in live edit mode
 3. make some obvious edits in your current language and save
 4. Observe that the page refreshes and displays your edits as expected

**To Test Locally**
This will help identify issues before deploying to dev
 1. Login in chrome in an **incognito** window.
 2. Open devtools > networking tab and go to the login page
 3. Observe that your network traffic (XHR) contains a request to /translation/[your_language_code] with a 302 response, followed by a second request to login.html with a 404 response (see the blue highlight)
 3. Login
 4. Observe that your latest network traffic looks like the green highlight - should be a successful second request to /translation/[your_langage_code]
5. Click on the latest request to /translation/[your_language_code] and verify that the request header contains the "router: 'api'" attribute.

![translationFetchNetwork](https://user-images.githubusercontent.com/49160279/109858119-6e3e1d80-7c29-11eb-82ea-9452c6301ede.png)


You can also try various combinations of different users, browsers, and deleting any of the browser cache for this site (cookies, local storage, session storage, indexeddb, etc.

**Please note the following:**
* Firefox incognito is not currently supported by indexeddb, and will cause the site to behave unpredictably.
* Deleting the site cookie data can force a logout
* Deleting indexeddb can trigger exceptions that require a page refresh as the browser finalizes the delete

With the exception of firefox incognito, the translations should behave as expected once the page has been stabilized with a refresh or logout.
